src/components/ColorPickerInput.tsx

```diff
@@ -0,0 +1,88 @@
+import {
+  useState,
+  useEffect,
+  useRef,
+  MutableRefObject,
+  useLayoutEffect,
+} from "react";
+import { Box, useTheme } from "@mui/material";
+
+import { Color, ColorPicker } from "react-color-palette";
+import { useDebouncedCallback } from "use-debounce";
+
+const useResponsiveWidth = (): [
+  width: number,
+  setRef: MutableRefObject<HTMLElement | null>
+] => {
+  const ref = useRef(null);
+  const [width, setWidth] = useState(0);
+
+  useLayoutEffect(() => {
+    if (!ref || !ref.current) {
+      return;
+    }
+    const resizeObserver = new ResizeObserver((targets) => {
+      const { width: currentWidth } = targets[0].contentRect;
+      setWidth(currentWidth);
+    });
+
+    resizeObserver.observe(ref.current);
+
+    return () => {
+      resizeObserver.disconnect();
+    };
+  }, []);
+
+  return [width, ref];
+};
+
+export interface IColorPickerProps {
+  value: Color;
+  handleOnChangeComplete: (color: Color) => void;
+  disabled?: boolean;
+}
+
+export default function ColorPickerInput({
+  value,
+  handleOnChangeComplete,
+  disabled = false,
+}: IColorPickerProps) {
+  const [localValue, setLocalValue] = useState(value);
+  const [width, setRef] = useResponsiveWidth();
+  const theme = useTheme();
+  const debouncedOnChangeComplete = useDebouncedCallback((color) => {
+    handleOnChangeComplete(color);
+  }, 100);
+
+  useEffect(() => {
+    debouncedOnChangeComplete(localValue);
+  }, [debouncedOnChangeComplete, localValue]);
```

Comment:
I’m not sure why this is done this way. react-color-picker already has a [`onChangeComplete` prop](https://github.com/Wondermarin/react-color-palette#colorpicker-props) fired on mouseup. Then you can avoid debouncing and using an effect.

If you need to debounce (when writing to db), it should be handled in the parent component. If the reason for debounce is to improve performance, you can use React 18’s [`startTransition`](https://reactjs.org/docs/react-api.html#starttransition)

```suggestion
```

Reply:
The problem here is every color changes frame(100 times+ just for a little drag) appends lots of styles into head as we're generating styles regarding. I'm aware of this problem for now. This also makes hard to reuse this component I already tried it for Color field. 

I'm actually planning to work on another related issue after this. I've decided to think through this reusability problems during slider customization implementation. I'll research how we can use startTransition btw.

Also, a similar implementation used in Color field's PopoverCell.

Reply:
> The problem here is every color changes frame(100 times+ just for a little drag)

It only re-renders this ColorPickerInput component, which is expected and necessary. This is why we use `onChangeComplete` to save the selected color only when the user has stopped interacting with the picker, and then update the other components like preview.

> appends lots of styles into head as we're generating styles regarding.

This only happens when the user stops dragging (`onChangeComplete`), and MUI only appends styles to head in development mode. Notice that in dev mode, there are 1000+ nodes of `<style data-emotion="mui" data-s>` but these do not appear in production (like demo.rowy.io). So this isn’t a problem.

> Also, a similar implementation used in Color field's PopoverCell.

This is because when `onChangeComplete` is called there, it immediately writes to the database, and we need to debounce writes for that since the user pays for writes. In this case, it’s all local, so we don’t have to add a debounce.

---

Additionally, now that you’ve accepted the suggestion to pass `onChangeComplete` directly, you’ve made this debounce + effect redundant. Adding logs, I saw that:
1. It fired when the dropdown is first opened.
2. It fired when the user stopped moving the mouse, but is still holding down the mouse button.
3. It fired when the user stopped holding the mouse button (this is the only time we want to fire), but since we already have an `onChangeComplete` prop, `onChangeComplete` is called twice at this point.

I’ve added another change request and when you click accept on that in GitHub, I’ll merge this PR.